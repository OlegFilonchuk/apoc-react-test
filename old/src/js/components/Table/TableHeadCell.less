.table-head-cell-with-sorting {
  cursor: pointer;

  .th-sorting-icon {
    display: inline-block;
    margin-right: 5px;
    top: 0;
    width: 10px;

    &.sc-icon-up-arrow,
    &.sc-icon-down-arrow {
      font-size: 7px;
      top: -3px;
      position: relative;

      &::before {
        padding-left: 3px;
        display: inline-block;
      }
    }
  }
}
