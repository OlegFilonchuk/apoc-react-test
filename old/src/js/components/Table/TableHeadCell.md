## Example of table sorting
It uses [Table](#table) and [TableHeadRow](#tableheadrow)

    function onSort(sortOrder, sortField) {
        setState({ sortOrder, sortField });
    }

    function getSortOrder(sortField) {
      return state.sortField === sortField ? state.sortOrder : ''
    }

    const initialState = {
      sortField: '',
      sortOrder: ''
    };

    (
      <Table>
        <TableHeadRow>
          <TableHeadCell sortOrder={getSortOrder('id-1')} sortable onSort={(newSortOrder) => onSort(newSortOrder, 'id-1')}>Title</TableHeadCell>
          <TableHeadCell sortOrder={getSortOrder('id-2')} sortable onSort={(newSortOrder) => onSort(newSortOrder, 'id-2')}>Title</TableHeadCell>
          <TableHeadCell sortOrder={getSortOrder('id-3')} sortable onSort={(newSortOrder) => onSort(newSortOrder, 'id-3')}>Title</TableHeadCell>
        </TableHeadRow>
        <tbody>
          <tr>
            <td>Content</td>
            <td>Content</td>
            <td>Content</td>
          </tr>
          <tr>
            <td>Content</td>
            <td>Content</td>
            <td>Content</td>
          </tr>
          <tr>
            <td>Content</td>
            <td>Content</td>
            <td>Content</td>
          </tr>
        </tbody>
      </Table>
    )

## Example of table sorting with default sort value
It uses [Table](#table) and [TableHeadRow](#tableheadrow)

    function onSort(sortOrder, sortField) {
        setState({ sortOrder, sortField });
    }

    function getSortOrder(sortField) {
      return state.sortField === sortField ? state.sortOrder : ''
    }

    const initialState = {
        sortField: 'id-2',
        sortOrder: 'asc'
      };

    (
      <Table>
        <TableHeadRow>
          <TableHeadCell sortOrder={getSortOrder('id-1')} sortable onSort={(newSortOrder) => onSort(newSortOrder, 'id-1')}>Title</TableHeadCell>
          <TableHeadCell sortOrder={getSortOrder('id-2')} sortable onSort={(newSortOrder) => onSort(newSortOrder, 'id-2')}>Title</TableHeadCell>
          <TableHeadCell sortOrder={getSortOrder('id-3')} sortable onSort={(newSortOrder) => onSort(newSortOrder, 'id-3')}>Title</TableHeadCell>
        </TableHeadRow>
        <tbody>
          <tr>
            <td>Content</td>
            <td>Content</td>
            <td>Content</td>
          </tr>
          <tr>
            <td>Content</td>
            <td>Content</td>
            <td>Content</td>
          </tr>
          <tr>
            <td>Content</td>
            <td>Content</td>
            <td>Content</td>
          </tr>
        </tbody>
      </Table>
    )

## Example of table with different aligned headers and content
It uses [Table](#table), [TableCell](#tablecell), [TableHeadRow](#tableheadrow) and [TableHeadCell](#tableheadcell)

    function onSort(sortOrder, sortField) {
        setState({ sortOrder, sortField });
    }

    function getSortOrder(sortField) {
      return state.sortField === sortField ? state.sortOrder : ''
    }

    const initialState = {
      sortField: '',
      sortOrder: ''
    };

    (
      <Table>
        <TableHeadRow>
          <TableHeadCell textContent sortOrder={getSortOrder('id-1')} sortable onSort={(newSortOrder) => onSort(newSortOrder, 'id-1')}>Text Content</TableHeadCell>
          <TableHeadCell iconOrGraphicalContent sortOrder={getSortOrder('id-2')} sortable onSort={(newSortOrder) => onSort(newSortOrder, 'id-2')}>Icon Content</TableHeadCell>
          <TableHeadCell numericContent sortOrder={getSortOrder('id-3')} sortable onSort={(newSortOrder) => onSort(newSortOrder, 'id-3')}>Numeric Content</TableHeadCell>
        </TableHeadRow>
        <tbody>
          <tr>
            <TableCell>Short text content</TableCell>
            <TableCell><span className="sc-icon-check sc-success" /></TableCell>
            <TableCell>$12.34</TableCell>
          </tr>
          <tr>
            <TableCell>Little longer text content</TableCell>
            <TableCell><span className="sc-grade-c sc-icon-triangle-warning" /></TableCell>
            <TableCell>$123.45</TableCell>
          </tr>
          <tr>
            <TableCell>Content</TableCell>
            <TableCell><span className="sc-icon-check sc-success" /></TableCell>
            <TableCell>$1234.56</TableCell>
          </tr>
        </tbody>
      </Table>
    )
