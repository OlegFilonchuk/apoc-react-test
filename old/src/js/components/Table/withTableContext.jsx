import React from 'react';
import propTypes from 'prop-types';

export const CONTEXT_ID = 'APOC_REACT_TABLE_CONTEXT';

export const tableContextPropTypes = propTypes.shape({
  /**
   * @type {function(uuid: string, index:number, contentTypes: CellContentTypes)}
   */
  registerColumn: propTypes.func,
  /**
   * @type {function(uuid: string)}
   */
  unregisterColumn: propTypes.func,
  /**
   * @type {function(index: ?number):CellContentTypes}
   */
  getColumnContentTypes: propTypes.func
});

/**
 * @type {CellContentTypes}
 */
export const defaultContentTypes = {
  dateContent: false,
  iconOrGraphicalContent: false,
  textContent: false,
  numericContent: false
};

export const defaultCellTableProps = {
  registerColumn: () => null,
  unregisterColumn: () => null,
  getColumnContentTypes: () => defaultContentTypes
};

export function withTableContext(Component) {
  function WithTableContext(props, context) {
    return <Component {...props} tableProps={context[CONTEXT_ID]} />;
  }

  WithTableContext.propTypes = Component.propTypes;

  WithTableContext.contextTypes = {
    [CONTEXT_ID]: tableContextPropTypes
  };

  WithTableContext.displayName = `withTableContext${Component.displayName || Component.name}`;

  WithTableContext.WrappedComponent = Component;

  return WithTableContext;
}

/* eslint-disable no-unused-vars */
/**
 * @typedef {{dateContent: boolean,iconOrGraphicalContent: boolean,textContent: boolean, numericContent: boolean}} CellContentTypes
 */
const CellContentTypes = {};
/* eslint-enable no-unused-vars */
