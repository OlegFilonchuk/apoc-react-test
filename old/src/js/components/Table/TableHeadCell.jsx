import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import uuid from 'uuid';

import { validateProps, isContentTypePassed } from './TableHelpers';
import { withTableContext, tableContextPropTypes, defaultCellTableProps } from './withTableContext';

import './TableHeadCell.less';

export const TABLE_SORT_ORDER = {
  NONE: '',
  ASC: 'asc',
  DESC: 'desc'
};

const SORT_QUEUE = {
  [TABLE_SORT_ORDER.NONE]: TABLE_SORT_ORDER.ASC,
  [TABLE_SORT_ORDER.ASC]: TABLE_SORT_ORDER.DESC,
  [TABLE_SORT_ORDER.DESC]: TABLE_SORT_ORDER.NONE
};

class TableHeadCell extends React.Component {
  componentDidMount() {
    this.uuid = uuid();

    this.props.tableProps.registerColumn(this.uuid, this.childIndex, {
      dateContent: this.props.dateContent,
      iconOrGraphicalContent: this.props.iconOrGraphicalContent,
      textContent: this.props.textContent,
      numericContent: this.props.numericContent
    });
  }

  componentWillUnmount() {
    this.props.tableProps.unregisterColumn(this.uuid);
  }

  onClick = () => {
    if (!this.props.sortable) {
      return null;
    }
    const newSortOrder = SORT_QUEUE[this.props.sortOrder];

    return this.props.onSort(newSortOrder);
  };

  setRef = ref => {
    this.ref = ref;
    this.forceUpdate();
  };

  get childIndex() {
    return Array.from(this.ref.parentElement.children).indexOf(this.ref);
  }

  render() {
    const {
      sortable,
      sortOrder,
      onSort,
      className,
      children,
      width,
      textContent,
      numericContent,
      iconOrGraphicalContent,
      dateContent,
      tableProps,
      ...props
    } = this.props;

    const thClassName = classNames(className, {
      'table-head-cell-with-sorting': sortable,
      'sc-left-align': textContent,
      'sc-right-align': numericContent,
      'sc-center-align': iconOrGraphicalContent || dateContent || !isContentTypePassed
    });
    const iconClassName = classNames('th-sorting-icon', className, {
      'sc-icon-dropdown-arrow': sortOrder === TABLE_SORT_ORDER.NONE,
      'sc-icon-up-arrow': sortOrder === TABLE_SORT_ORDER.ASC,
      'sc-icon-down-arrow': sortOrder === TABLE_SORT_ORDER.DESC
    });
    const style = width
      ? {
          width,
          maxWidth: width
        }
      : {};

    return (
      <th className={thClassName} onClick={this.onClick} style={style} {...props} ref={this.setRef}>
        {sortable && <span className={iconClassName} />}
        {children}
      </th>
    );
  }
}

TableHeadCell.propTypes = {
  /**
   * Cell's content
   */
  children: PropTypes.node,

  /**
   * List of custom classes added to component
   */
  className: PropTypes.string,

  /**
   * Show if cell has possibility to sort
   */
  sortable: PropTypes.bool,

  /**
   * It called when cell is clicked
   */
  onSort: PropTypes.func,

  /**
   * Current order for sorting table cell
   */
  sortOrder: PropTypes.string,

  /**
   * Cell's width
   */
  width: PropTypes.string,

  /**
   * shows if cell's content is dates with consistent character count
   */
  dateContent: validateProps,

  /**
   * shows if cell's content is icons or other individual Graphical elements with no Label Accompaniment
   */
  iconOrGraphicalContent: validateProps,

  /**
   * shows if cell's content is text strings
   */
  textContent: validateProps,

  /**
   * shows if cell's content is numeric values with inconsistent character counts (e.g. money, file sizes, byte sizes etc.)
   */
  numericContent: validateProps,

  tableProps: tableContextPropTypes
};

TableHeadCell.defaultProps = {
  sortable: false,
  sortOrder: '',
  className: '',
  children: null,
  onSort: () => null,
  width: '',
  dateContent: false,
  iconOrGraphicalContent: false,
  textContent: false,
  numericContent: false,
  tableProps: defaultCellTableProps
};

const TableHeadWithContext = withTableContext(TableHeadCell);

export default TableHeadWithContext;
