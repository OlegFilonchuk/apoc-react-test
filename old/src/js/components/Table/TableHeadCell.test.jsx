import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import TableHeadCell, { TABLE_SORT_ORDER } from './TableHeadCell';
import { defaultContentTypes } from './withTableContext';

const tr = document.createElement('tr');

const mountCell = (props = {}) =>
  mount(<TableHeadCell.WrappedComponent {...props}>Title</TableHeadCell.WrappedComponent>, {
    attachTo: tr
  });

describe('<TableHeadCell />', () => {
  const tableProps = {
    registerColumn: jest.fn(),
    unregisterColumn: jest.fn(),
    getColumnContentTypes: jest.fn()
  };

  describe('should match snapshot ', () => {
    it('with passing `width`', () => {
      expect(toJson(mountCell({ width: '100px' }))).toMatchSnapshot();
    });

    it('without passing `width`', () => {
      expect(toJson(mountCell())).toMatchSnapshot();
    });

    it('with sorting', () => {
      expect(toJson(mountCell({ sortable: true }))).toMatchSnapshot();
    });
  });

  describe('set `className` for icon correctly when `sortOrder` is', () => {
    it(TABLE_SORT_ORDER.NONE, () => {
      const cell = mountCell({ sortable: true, sortOrder: TABLE_SORT_ORDER.NONE });

      expect(cell.find('.th-sorting-icon')).toHaveClassName('sc-icon-dropdown-arrow');
    });

    it(TABLE_SORT_ORDER.ASC, () => {
      const cell = mountCell({ sortable: true, sortOrder: TABLE_SORT_ORDER.ASC });

      expect(cell.find('.th-sorting-icon')).toHaveClassName('sc-icon-up-arrow');
    });

    it(TABLE_SORT_ORDER.DESC, () => {
      const cell = mountCell({ sortable: true, sortOrder: TABLE_SORT_ORDER.DESC });

      expect(cell.find('.th-sorting-icon')).toHaveClassName('sc-icon-down-arrow');
    });
  });

  it('correctly set `className`', () => {
    const cell = mountCell({ className: 'some-extra-class' });
    const element = cell.find('th');

    expect(element).toHaveClassName('some-extra-class');
  });

  it('correctly set `width`', () => {
    const cell = mountCell({ width: '100px' });
    const element = cell.find('th');

    expect(element).toHaveStyle('width', '100px');
    expect(element).toHaveStyle('maxWidth', '100px');
  });

  it('`onSort` works correctly', () => {
    let sortOrder = TABLE_SORT_ORDER.NONE;

    function onSort(newSortOrder) {
      sortOrder = newSortOrder;
    }

    const cell = mountCell({ sortable: true, sortOrder, onSort });

    cell.simulate('click');
    expect(sortOrder).toEqual(TABLE_SORT_ORDER.ASC);
    cell.setProps({ sortOrder });
    cell.simulate('click');
    expect(sortOrder).toEqual(TABLE_SORT_ORDER.DESC);
    cell.setProps({ sortOrder });
    cell.simulate('click');
    expect(sortOrder).toEqual(TABLE_SORT_ORDER.NONE);
  });

  describe('without contentTypes set', () => {
    let theCell;

    beforeAll(() => {
      tableProps.registerColumn.mockClear();
      theCell = mountCell({ tableProps });
    });
    describe('onComponentDidMount', () => {
      it('calls the tableProps.registerColumn', () => {
        expect(theCell.props({}).tableProps).toBe(tableProps);
        expect(tableProps.registerColumn).toHaveBeenCalledWith(theCell.instance().uuid, 0, defaultContentTypes);
      });
    });
  });

  describe('with contentTypes set', () => {
    let theCell;

    beforeAll(() => {
      tableProps.registerColumn.mockClear();
      theCell = mountCell({ tableProps, textContent: true });
    });
    describe('onComponentDidMount', () => {
      it('calls the tableProps.registerColumn', () => {
        expect(theCell.props({}).tableProps).toBe(tableProps);
        expect(tableProps.registerColumn).toHaveBeenCalledWith(theCell.instance().uuid, 0, {
          dateContent: false,
          iconOrGraphicalContent: false,
          textContent: true,
          numericContent: false
        });
      });
    });
  });
});
