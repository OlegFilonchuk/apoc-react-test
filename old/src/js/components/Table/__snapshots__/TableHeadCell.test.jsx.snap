// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableHeadCell /> should match snapshot  with passing \`width\` 1`] = `
<TableHeadCell
  className=""
  dateContent={false}
  iconOrGraphicalContent={false}
  numericContent={false}
  onSort={[Function]}
  sortOrder=""
  sortable={false}
  tableProps={
    Object {
      "getColumnContentTypes": [Function],
      "registerColumn": [Function],
      "unregisterColumn": [Function],
    }
  }
  textContent={false}
  width="100px"
>
  <th
    className=""
    onClick={[Function]}
    style={
      Object {
        "maxWidth": "100px",
        "width": "100px",
      }
    }
  >
    Title
  </th>
</TableHeadCell>
`;

exports[`<TableHeadCell /> should match snapshot  with sorting 1`] = `
<TableHeadCell
  className=""
  dateContent={false}
  iconOrGraphicalContent={false}
  numericContent={false}
  onSort={[Function]}
  sortOrder=""
  sortable={true}
  tableProps={
    Object {
      "getColumnContentTypes": [Function],
      "registerColumn": [Function],
      "unregisterColumn": [Function],
    }
  }
  textContent={false}
  width=""
>
  <th
    className="table-head-cell-with-sorting"
    onClick={[Function]}
    style={Object {}}
  >
    <span
      className="th-sorting-icon sc-icon-dropdown-arrow"
    />
    Title
  </th>
</TableHeadCell>
`;

exports[`<TableHeadCell /> should match snapshot  without passing \`width\` 1`] = `
<TableHeadCell
  className=""
  dateContent={false}
  iconOrGraphicalContent={false}
  numericContent={false}
  onSort={[Function]}
  sortOrder=""
  sortable={false}
  tableProps={
    Object {
      "getColumnContentTypes": [Function],
      "registerColumn": [Function],
      "unregisterColumn": [Function],
    }
  }
  textContent={false}
  width=""
>
  <th
    className=""
    onClick={[Function]}
    style={Object {}}
  >
    Title
  </th>
</TableHeadCell>
`;
